tests/test_pd_controller.cpp

```diff
@@ -0,0 +1,226 @@
+constexpr auto MakeArray(T &&...values) -> std::array<ValueType, sizeof...(T)> {
+  return std::array<ValueType, sizeof...(T)>{std::forward<T>(values)...};
+}
+
+// RAII mutator, use to temporary modify a value given a reference and a tmp
+// value
+//
+// WARNING: Dangling references & co ...
+template <typename ValueType>
+struct TemporaryMutate {
+  TemporaryMutate() = delete;
+
+  template <typename T>
+  TemporaryMutate(ValueType &data, T &&tmp_value)
+      : data_(data), before_(data_) {
+    data_ = std::forward<T>(tmp_value);
+  }
+
+  virtual ~TemporaryMutate() { data_ = before_; }
+
+  constexpr auto GetOldValue() const -> const ValueType & { return before_; }
+
+ private:
+  ValueType &data_;
+  ValueType before_;
+};
+
+TEST(PdControllerTest, Ctor) {
+  EXPECT_NO_THROW({ const auto pd_ctrl = PDController(); });
+}
+
+struct Gains {
+  using IdxType = Eigen::Index;
+
+  Eigen::VectorXd p;
+  Eigen::VectorXd d;
+
+  static auto Random(IdxType p_size,
+                     std::optional<IdxType> d_size = std::nullopt) -> Gains {
+    return Gains{
+        .p = Eigen::VectorXd::Random(p_size),
+        .d = Eigen::VectorXd::Random(d_size.value_or(p_size)),
+    };
+  }
+};
+
+TEST(PdControllerTest, SetGains) {
+  auto pd_ctrl = PDController();
+
+  EXPECT_NO_THROW({ pd_ctrl.set_gains(Eigen::VectorXd{}, Eigen::VectorXd{}); });
+
+  for (const auto size : MakeArray<Gains::IdxType>(1, 3, 4, 50, 1000)) {
+    SCOPED_TRACE(::testing::Message() << "Same gains size: " << size);
+    auto requested_gains = Gains::Random(size);
+    EXPECT_NO_THROW(
+        { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+    // FIXME: How to test the validity ??
+
+    // Test for special double values acceptance
+    for (const auto special_value : {
+             std::numeric_limits<double>::infinity(),
+             std::numeric_limits<double>::quiet_NaN(),
+             std::numeric_limits<double>::signaling_NaN(),
+         }) {
+      SCOPED_TRACE(::testing::Message() << "Special value: " << special_value);
+
+      {
+        SCOPED_TRACE("P");
+        auto _ = TemporaryMutate{requested_gains.p(0), special_value};
+        EXPECT_NO_THROW(
+            { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+      }
+
+      {
+        SCOPED_TRACE("D");
+        auto _ = TemporaryMutate{requested_gains.d(0), special_value};
+        EXPECT_NO_THROW(
+            { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+      }
+    }
+  }
+
+  using SizePair = std::array<Gains::IdxType, 2>;
+  for (const auto [p_size, d_size] : MakeArray<SizePair>(
+           SizePair{1, 2}, SizePair{4, 3}, SizePair{50, 1000})) {
+    SCOPED_TRACE(::testing::Message()
+                 << "Different Sizes: P = " << p_size << ", D = " << d_size);
+    const auto requested_gains = Gains::Random(p_size, d_size);
+    // FIXME: I guess it should fail but set_gains doesn't provide any feedback
+    // on failure
```

Comment:
I agree we should fail in this case!

Reply:
No. Using pinocchio for this is an overkill